[조이] 뉴스스탠드 2주차

youzysu opened the pull request:
![Screen Recording 2023-06-03 at 23 03 58](https://github.com/codesquad-members-2023/fe-max--newsstand/assets/111998760/e83239f1-a47d-4c98-b6bd-e58774f4f198)

# 2주차 주요 내용
> [리드미](https://github.com/codesquad-members-2023/fe-max--newsstand/pull/23/commits/2d5fa7434affa486625f28489898ab1bf8b5bf68)

youzysu committed:
feat #12, #13: 전체 언론사 그리드 영역 좌우 화살표로 언론사 페이지 넘기는 기능

- 좌우에 화살표로 언론사 페이지를 넘긴다.
- startIndex를 props로 전달받아 가장 첫 페이지의 왼쪽 화살표, 끝 페이지의 오른쪽 화살표는 표시되지 않는다.

youzysu committed:
feat #11, #20: 언론사 브랜드 마크 순서 랜덤 배치하기, 로고 클릭 시 화면 새로고침하기

- click eventhandler
- shuffleArray 유틸 함수

youzysu committed:
feat #18: 언론사 구독 리스트 초기 상태 만들기

- 각 언론사 별로 구독 상태를 boolean 값으로 가진다.
- 전체 언론사 이름을 key로, 구독 상태를 value로 하는 구독 리스트를 받아온다.

Kakamotobi commented on src/store/index.ts:
```diff
@@ -15,6 +15,14 @@ const initialState: NewsStandState = {
+    case 'MOVE_GRID': {
+      const { direction } = action.payload;
+      const nextStartIndex =
+        direction === 'left' ? state.gridPressStartIndex - 24 : state.gridPressStartIndex + 24;
```
6, 4, 24 숫자들이 여기저기서 자주 등장하는데 변수에 할당해주시면 나중에 유지보수하기 더 좋을거 같아요.

Kakamotobi commented on src/components/MediaArea/GridViewer/GridButton.ts:
```diff
@@ -0,0 +1,51 @@
+
+interface GridButtonProps {
+  startIndex: number;
+  type: 'left' | 'right';
```
밑에 21번줄에서 이름을 다시 정해주고 있는데 애초에 "type"말고 "direction"으로 정의를 해두면 더 일관성 있고 읽기 편할것 같아요.

crongro commented:
전체적으로 깔끔하네요.
foreach이외 다른 고차함수도 많이 활용하시고요. 
에러처리도 한번 신경써보세요.
생성자는 지금보다 더 커지지 않도록 하고요.
rAF활용도 좋네요.

Kakamotobi commented on server/index.ts:
```diff
@@ -1,15 +1,34 @@
-app.use(cors());
+app.use(cors({ origin: 'http://localhost:5173' }));
 
 app.get('/', (_, res) => res.status(200).send('Hello from express!'));
```
해당 endpoint는 인사해줘서 기분은 좋지만 따로 serve 해주는게 없어서 불필요해보입니다.

namse replied:
Health 체크 용으로 내비둬도 될듯

Kakamotobi commented on src/components/MediaArea/GridViewer/GridViewer.ts:
```diff
@@ -9,36 +9,50 @@ interface GridViewerProps {
 
   private setProps() {
     const { pressList, startIndex } = this.props;
-    const { gridRows, grids } = this;
+    const { grids, gridIcons } = this;
```
가독성은 개인마다 차이가 있겠지만 `const { props: { grids, gridIcons }, pressList, startIndex } = this.props;` 처럼 destructuring해도 될것 같습니다.

namse commented on src/store/index.ts:
```diff
@@ -1,20 +1,29 @@
   systemDate: new Date(),
   trendNewsList: await fetchNewsList(),
-  allPressList: await fetchPressList(),
+  allPressList: shuffleArray(await fetchPressList()),
   leftNewsIndex: 0,
   rightNewsIndex: 1,
   tabOption: 'all',
   viewerOption: 'grid',
   gridPressStartIndex: 0,
+  subscribePressList: await fetchSubscribePressList(),
```
제안하고 싶은 것: 처음엔 빈 배열로 먼저 렌더링하고, fetch가 오면 그때 state를 업데이트해서 재렌더링해보세요.

youzysu committed:
refactor #4: 전체 언론사 그리드 보기 컴포넌트 로직 개선

- 프로퍼티 업데이트 시 기존 언론사 아이콘 모두 지운 후 새로운 상태 넣어주기

youzysu committed:
feat #21: Action interface 개선

- 액션 타입에 따라 type, payload 인터페이스를 만들어준다.

youzysu committed:
feat: 언론사 구독 리스트 JSON 파일 생성

youzysu committed:
refactor #5: GridViewer 테이블 만드는 로직 Grid 개수 매직넘버 상수 분리

youzysu committed:
refactor: 가독성을 위한 언론사 그리드 데이터 수정

- key: src -> icon
- value: alt -> name

youzysu committed:
feat #22: 언론사 그리드에 마우스를 올리면 구독 상태에 따라 구독 버튼 보여주기

- GridViewer 하위 컴포넌트 생성: Grid 컴포넌트 분리
- 재사용성을 위해 SubscribeButton 컴포넌트 분리
- 구독 언론사 리스트를 props로 전달한다.

youzysu committed:
feat #18, #22: 언론사 구독 해지 기능, 구독 상태 변경 시 리렌더링

- 구독하기 버튼 클릭 시 해당 언론사의 구독 상태를 변경한다.
- GridViewer 업데이트 로직 반영

youzysu committed:
refactor: 뉴스바 롤링 영역 리팩터링

- Store에서 index % length index 조정해준다.
- 단, nextHeadline의 경우 currentIndex가 마지막 요소인 경우 여전히 나머지로 해줘야 한다.

youzysu committed:
refactor: Action interface 추가 생성 및 네이밍 수정

youzysu committed:
refactor #22: 마우스 올린 경우에만 구독 버튼 Element 추가되도록 리팩터링

- 불필요하게 미리 Element를 DOM에 추가하지 않는다.

youzysu committed:
refactor #21, #24: 초기 상태와 data fetch 로직을 분리하여 DOM 생성 이후 데이터로 재렌더링한다.

- 컴포넌트 생성 시 빈값으로 Element를 모두 생성한다.
- componentDidMount: DOM 생성 이후 초기화에 필요한 로직을 진행한다.
- thunkDispatch: 데이터 fetch 작업을 나중에 하도록 함수로 감싸서 Action을 실행한다.
- 언론사 구독 리스트를 로컬 스토리지와 연동(동기화)한다.

youzysu committed:
docs #25: sprint 02 작업 내용 리드미 작성

jsh3418 commented on src/components/MediaArea/MediaArea.ts:
```diff
@@ -0,0 +1,54 @@
+  constructor() {
+    this.element = createElement('DIV', { class: styles.mediaArea });
+    this.gridViewer = new GridViewer();
+    this.leftButton = new GridButton({ type: 'left' });
+    this.rightButton = new GridButton({ type: 'right' });
```
객체로 넘기도록 하신 이유가 궁금합니다! 🤔 받는 쪽에서 바로 분해 할당하는 것 같은데 굳이? 라는 생각이라서

youzysu replied:
'left', 'right'가 의미하는걸 좀 더 명확하게 표현하고 싶었어요 ㅎㅎ

jsh3418 commented on src/components/MediaArea/GridViewer/GridButton.ts:
```diff
@@ -0,0 +1,41 @@
+  private element;
+  private type: 'left' | 'right';
+
+  constructor({ type: type }: { type: 'left' | 'right' }) {
```
```ts
constructor({ type }: { type: 'left' | 'right' }) {
```
타입 안써도 될 것 같습니다.

jsh3418 replied:
```type: 'left' | 'right'```가 반복되고 있어서 타입을 지정해주면 좋을 것 같아요

jsh3418 commented on src/components/MediaArea/GridViewer/GridButton.ts:
```diff
@@ -0,0 +1,41 @@
+
+  public render({ startIndex }: GridButtonProps) {
+    const isFirstPage = startIndex === 0;
+    const isLastPage = startIndex === 72;
```
J: 72?

jsh3418 commented on src/components/MediaArea/GridViewer/GridButton.ts:
```diff
@@ -0,0 +1,41 @@
+  startIndex: number;
+}
+
+export default class GridButton {
```
bakha: `GridButton`은 너무 추상적입니다. 좀 더 명확한 클래스 이름이 필요하다고 생각합니다.

youzysu replied:
Grid에 버튼이 있다고 생각할 수 있겠네요~ GridViewerButton으로 수정할게요
